File: clients/python/src/masugate_client/operation_contract.py
"""Public validators for the additive governed-route-manifest v2 contract."""

from __future__ import annotations

from collections.abc import Mapping
from typing import TypeAlias, cast

from .adapter_contract import (
    GOVERNED_ROUTE_MANIFEST_VERSION,
    GovernedRouteManifest,
    _canonical_json,
    _identifier,
    _record,
    require_adapter_argument_name,
    validate_governed_route_manifest,
)
from .models import JsonValue

GOVERNED_ROUTE_MANIFEST_V2_VERSION = "masugate.governed-route-manifest.v2"
DEFAULT_ROUTE_SCHEMA_CANONICAL_BYTES = 65_536
DEFAULT_ROUTE_MANIFEST_CANONICAL_BYTES = 1_048_576
MAX_GOVERNED_ROUTE_MANIFEST_ROUTES = 64
MAX_ROUTE_ARTIFACT_FIELDS = 128
MAX_ROUTE_CONNECTOR_CAPABILITIES = 64
GovernedRouteManifestV2: TypeAlias = dict[str, JsonValue]
AnyGovernedRouteManifest: TypeAlias = GovernedRouteManifest | GovernedRouteManifestV2
_SECRET_MODEL_FIELD_PARTS = frozenset(
    {
        "credential",
        "credentials",
        "secret",
        "secrets",
        "token",
        "tokens",
        "password",
        "apikey",
        "privatekey",
        "accesskey",
    }
)
_SECRET_MODEL_FIELD_COMPOUNDS = frozenset({"api_key", "private_key", "access_key"})


def _model_field(value: object, field: str) -> str:
    if type(value) is not str:
        raise ValueError(f"{field} must be a model field")
    parsed = require_adapter_argument_name(value)
    if parsed == "runtime" or parsed.startswith("model_"):
        raise ValueError(f"{field} uses a reserved generated-host name")
    parts = parsed.split("_")
    if any(part in _SECRET_MODEL_FIELD_PARTS for part in parts) or any(
        "_".join(parts[index : index + 2]) in _SECRET_MODEL_FIELD_COMPOUNDS
        for index in range(len(parts) - 1)
    ):
        raise ValueError(f"{field} cannot name secret or credential material")
    return parsed


def _exact(value: Mapping[str, object], allowed: set[str], field: str) -> None:
    if set(value) != allowed:
        raise ValueError(f"{field} must contain exactly: {', '.join(sorted(allowed))}")


def _digest(value: object, field: str) -> str:
    if (
        type(value) is not str
        or len(value) != 64
        or any(c not in "0123456789abcdef" for c in value)
    ):
        raise ValueError(f"{field} must be a lowercase SHA-256 digest")
    return value


def _integer(value: object, field: str, minimum: int, maximum: int) -> int:
    if type(value) is not int or not minimum <= value <= maximum:
        raise ValueError(f"{field} has invalid integer bounds")
    return value


def _schema_record(value: object, field: str, *, max_entries: int) -> Mapping[str, object]:
    if not isinstance(value, Mapping):
        raise ValueError(f"{field} must be an object with string keys")
    if len(value) > max_entries:
        raise ValueError(f"{field} must contain at most {max_entries} entries")
    if not all(type(key) is str for key in value):
        raise ValueError(f"{field} must be an object with string keys")
    return cast(Mapping[str, object], value)


def _canonical_bytes(value: JsonValue) -> int:
    return len(_canonical_json(value).encode("utf-8"))


class _SchemaByteBudget:
    """Incrementally account for the canonical bytes of one schema tree."""

    def __init__(self, limit: int, field: str) -> None:
        self.remaining = limit
        self.field = field

    def consume(self, amount: int) -> None:
        self.remaining -= amount
        if self.remaining < 0:
            raise ValueError(f"{self.field} canonical form exceeds configured limit")


def _schema(
    value: object, field: str, *, limit: int, object_root: bool = False
) -> dict[str, JsonValue]:
    budget = _SchemaByteBudget(limit, field)

    def visit(raw: object, context: str, depth: int) -> dict[str, JsonValue]:
        if depth > 8:
            raise ValueError(f"{context} exceeds maximum schema nesting")
        record = _schema_record(raw, context, max_entries=5)
        kind = record.get("type")
        if kind == "object":
            _exact(record, {"type", "properties", "required", "additionalProperties"}, context)
            if record["additionalProperties"] is not False:
                raise ValueError(f"{context}.additionalProperties must be false")
            properties = _schema_record(
                record["properties"], f"{context}.properties", max_entries=128
            )
            if not properties or len(properties) > 128:
                raise ValueError(f"{context}.properties must contain 1 through 128 fields")
            property_items = [
                (_model_field(name, f"{context}.properties key"), child)
                for name, child in properties.items()
            ]
            property_names = {name for name, _child in property_items}
            required_raw = record["required"]
            if not isinstance(required_raw, list | tuple):
                raise ValueError(f"{context}.required must be an array")
            if len(required_raw) > len(property_items):
                raise ValueError(f"{context}.required must name unique declared properties")
            required: list[str] = []
            for index, name in enumerate(required_raw):
                parsed_name = _model_field(name, f"{context}.required[{index}]")
                if parsed_name not in property_names or parsed_name in required:
                    raise ValueError(f"{context}.required must name unique declared properties")
                required.append(parsed_name)
            required.sort()
            object_shell: dict[str, JsonValue] = {
                "type": "object",
                "properties": {},
                "required": cast(JsonValue, required),
                "additionalProperties": False,
            }
            budget.consume(
                _canonical_bytes(object_shell)
                + sum(_canonical_bytes(name) + 1 for name, _child in property_items)
                + len(property_items)
                - 1
            )
            parsed_properties: dict[str, JsonValue] = {
                name: visit(child, f"{context}.properties.{name}", depth + 1)
                for name, child in property_items
            }
            return {
                "type": "object",
                "properties": parsed_properties,
                "required": cast(JsonValue, required),
                "additionalProperties": False,
            }
        if kind == "array":
            allowed = {"type", "items", "minItems", "maxItems"}
            if set(record) - allowed or "items" not in record or "maxItems" not in record:
                raise ValueError(f"{context} array schemas must be explicitly bounded")
            maximum = _integer(record["maxItems"], f"{context}.maxItems", 0, 1024)
            minimum = _integer(record.get("minItems", 0), f"{context}.minItems", 0, maximum)
            array_shell: dict[str, JsonValue] = {
                "type": "array",
                "items": None,
                "minItems": minimum,
                "maxItems": maximum,
            }
            budget.consume(_canonical_bytes(array_shell) - len("null"))
            return {
                "type": "array",
                "items": visit(record["items"], f"{context}.items", depth + 1),
                "minItems": minimum,
                "maxItems": maximum,
            }
        if kind == "string":
            if set(record) - {"type", "minLength", "maxLength"} or "maxLength" not in record:
                raise ValueError(f"{context} strings must declare maxLength")
            maximum = _integer(record["maxLength"], f"{context}.maxLength", 0, 65_536)
            minimum = _integer(record.get("minLength", 0), f"{context}.minLength", 0, maximum)
            string_schema: dict[str, JsonValue] = {
                "type": "string",
                "minLength": minimum,
                "maxLength": maximum,
            }
            budget.consume(_canonical_bytes(string_schema))
            return string_schema
        if kind == "integer":
            if set(record) - {"type", "minimum", "maximum"}:
                raise ValueError(f"{context} integers permit only safe bounds")
            minimum = _integer(
                record.get("minimum", -9_007_199_254_740_991),
                f"{context}.minimum",
                -9_007_199_254_740_991,
                9_007_199_254_740_991,
            )
            maximum = _integer(
                record.get("maximum", 9_007_199_254_740_991),
                f"{context}.maximum",
                minimum,
                9_007_199_254_740_991,
            )
            integer_schema: dict[str, JsonValue] = {
                "type": "integer",
                "minimum": minimum,
                "maximum": maximum,
            }
            budget.consume(_canonical_bytes(integer_schema))
            return integer_schema
        if kind == "boolean" and set(record) == {"type"}:
            boolean_schema: dict[str, JsonValue] = {"type": "boolean"}
            budget.consume(_canonical_bytes(boolean_schema))
            return boolean_schema
        raise ValueError(f"{context}.type is unsupported")

    parsed = visit(value, field, 0)
    if object_root and parsed["type"] != "object":
        raise ValueError(f"{field} must be an object schema")
    if len(_canonical_json(parsed).encode("utf-8")) > limit:
        raise ValueError(f"{field} canonical form exceeds configured limit")
    return parsed


def validate_governed_route_manifest_v2(
    value: object,
    *,
    max_schema_canonical_bytes: int = DEFAULT_ROUTE_SCHEMA_CANONICAL_BYTES,
    max_manifest_canonical_bytes: int = DEFAULT_ROUTE_MANIFEST_CANONICAL_BYTES,
) -> GovernedRouteManifestV2:
    """Validate a public v2 route projection without accepting deployment secrets."""

    if type(max_schema_canonical_bytes) is not int or max_schema_canonical_bytes <= 0:
        raise ValueError("schema canonical byte limit must be positive")
    if type(max_manifest_canonical_bytes) is not int or max_manifest_canonical_bytes <= 0:
        raise ValueError("route manifest canonical byte limit must be positive")
    manifest = _record(value, "governed route manifest v2")
    _exact(manifest, {"contract_version", "pack", "routes"}, "governed route manifest v2")
    if manifest["contract_version"] != GOVERNED_ROUTE_MANIFEST_V2_VERSION:
        raise ValueError("governed route manifest v2.contract_version is unsupported")
    pack = _record(manifest["pack"], "governed route manifest v2.pack")
    _exact(pack, {"id", "version", "digest"}, "governed route manifest v2.pack")
    parsed_pack: dict[str, JsonValue] = {
        "id": _identifier(pack["id"], "governed route manifest v2.pack.id"),
        "version": _identifier(pack["version"], "governed route manifest v2.pack.version"),
        "digest": _digest(pack["digest"], "governed route manifest v2.pack.digest"),
    }
    raw_routes = manifest["routes"]
    if not isinstance(raw_routes, list | tuple) or not raw_routes:
        raise ValueError("governed route manifest v2.routes must be a non-empty array")
    if len(raw_routes) > MAX_GOVERNED_ROUTE_MANIFEST_ROUTES:
        raise ValueError(
            "governed route manifest v2.routes must contain at most "
            f"{MAX_GOVERNED_ROUTE_MANIFEST_ROUTES} entries"
        )
    routes: list[dict[str, JsonValue]] = []
    seen_host_tools: set[str] = set()
    seen_actions: set[str] = set()
    for index, raw_route in enumerate(raw_routes):
        field = f"governed route manifest v2.routes[{index}]"
        route = _record(raw_route, field)
        _exact(
            route,
            {
                "host_tool",
                "action",
                "input_schema",
                "public_result_schema",
                "artifact_fields",
                "owner",
                "required_connector_capabilities",
                "maturity",
                "compatibility",
            },
            field,
        )
        host_tool = _identifier(route["host_tool"], f"{field}.host_tool")
        if host_tool in seen_host_tools:
            raise ValueError("governed route manifest v2.routes must not repeat host_tool")
        seen_host_tools.add(host_tool)
        action = _identifier(route["action"], f"{field}.action", max_length=255)
        if action in seen_actions:
            raise ValueError("governed route manifest v2.routes must not repeat action")
        seen_actions.add(action)
        owner_raw = _record(route["owner"], f"{field}.owner")
        position = owner_raw.get("position")
        if position == "transactional":
            _exact(owner_raw, {"provider_id", "position"}, f"{field}.owner")
            owner: dict[str, JsonValue] = {
                "provider_id": _identifier(owner_raw["provider_id"], f"{field}.owner.provider_id"),
                "position": "transactional",
            }
        elif position == "protected-external":
            _exact(owner_raw, {"provider_id", "position", "connector_id"}, f"{field}.owner")
            owner = {
                "provider_id": _identifier(owner_raw["provider_id"], f"{field}.owner.provider_id"),
                "position": "protected-external",
                "connector_id": _identifier(
                    owner_raw["connector_id"], f"{field}.owner.connector_id"
                ),
            }
        else:
            raise ValueError(f"{field}.owner.position is invalid")
        artifact_raw = route["artifact_fields"]
        capabilities_raw = route["required_connector_capabilities"]
        if not isinstance(artifact_raw, list | tuple) or not isinstance(
            capabilities_raw, list | tuple
        ):
            raise ValueError(f"{field} artifact fields and capabilities must be arrays")
        if len(artifact_raw) > MAX_ROUTE_ARTIFACT_FIELDS:
            raise ValueError(
                f"{field}.artifact_fields must contain at most {MAX_ROUTE_ARTIFACT_FIELDS} entries"
            )
        if len(capabilities_raw) > MAX_ROUTE_CONNECTOR_CAPABILITIES:
            raise ValueError(
                f"{field}.required_connector_capabilities must contain at most "
                f"{MAX_ROUTE_CONNECTOR_CAPABILITIES} entries"
            )
        artifacts = [_model_field(name, f"{field}.artifact_fields") for name in artifact_raw]
        capabilities = [
            _identifier(name, f"{field}.required_connector_capabilities")
            for name in capabilities_raw
        ]
        if len(set(artifacts)) != len(artifacts) or len(set(capabilities)) != len(capabilities):
            raise ValueError(f"{field} arrays must not contain duplicates")
        if position == "transactional" and capabilities:
            raise ValueError(f"{field} transactional route cannot require connector capabilities")
        maturity = route["maturity"]
        if maturity not in ("reference-effect", "production-profile"):
            raise ValueError(f"{field}.maturity is invalid")
        if maturity == "production-profile" and position != "protected-external":
            raise ValueError(f"{field}.production-profile requires protected-external position")
        compatibility = _record(route["compatibility"], f"{field}.compatibility")
        _exact(compatibility, {"route_manifest", "connector_contract"}, f"{field}.compatibility")
        if (
            compatibility["route_manifest"] != GOVERNED_ROUTE_MANIFEST_V2_VERSION
            or compatibility["connector_contract"] != "masugate.connector.v1"
        ):
            raise ValueError(f"{field}.compatibility is unsupported")
        input_schema = _schema(
            route["input_schema"],
            f"{field}.input_schema",
            limit=max_schema_canonical_bytes,
            object_root=True,
        )
        properties = cast(dict[str, JsonValue], input_schema["properties"])
        if any(name not in properties for name in artifacts):
            raise ValueError(f"{field}.artifact_fields must name input properties")
        if artifacts:
            if position != "protected-external":
                raise ValueError(f"{field}.artifact_fields require protected-external position")
            required = cast(list[str], input_schema["required"])
            for artifact in artifacts:
                property_schema = cast(dict[str, JsonValue], properties[artifact])
                if artifact not in required or property_schema.get("type") != "string":
                    raise ValueError(
                        f"{field}.artifact_fields must be required bounded string properties"
                    )
        routes.append(
            {
                "host_tool": host_tool,
                "action": action,
                "input_schema": input_schema,
                "public_result_schema": _schema(
                    route["public_result_schema"],
                    f"{field}.public_result_schema",
                    limit=max_schema_canonical_bytes,
                ),
                "artifact_fields": cast(JsonValue, artifacts),
                "owner": owner,
                "required_connector_capabilities": cast(JsonValue, capabilities),
                "maturity": cast(JsonValue, maturity),
                "compatibility": {
                    "route_manifest": GOVERNED_ROUTE_MANIFEST_V2_VERSION,
                    "connector_contract": "masugate.connector.v1",
                },
            }
        )
    routes.sort(key=lambda item: cast(str, item["host_tool"]))
    parsed: GovernedRouteManifestV2 = {
        "contract_version": GOVERNED_ROUTE_MANIFEST_V2_VERSION,
        "pack": parsed_pack,
        "routes": cast(JsonValue, routes),
    }
    if len(_canonical_json(parsed).encode("utf-8")) > max_manifest_canonical_bytes:
        raise ValueError("governed route manifest v2 canonical form exceeds configured limit")
    return parsed


def validate_any_governed_route_manifest(value: object) -> AnyGovernedRouteManifest:
    if (
        isinstance(value, Mapping)
        and value.get("contract_version") == GOVERNED_ROUTE_MANIFEST_VERSION
    ):
        return validate_governed_route_manifest(value)
    return validate_governed_route_manifest_v2(value)


def canonical_governed_route_manifest_v2(value: object) -> str:
    return _canonical_json(validate_governed_route_manifest_v2(value))


def canonical_any_governed_route_manifest(value: object) -> str:
    if (
        isinstance(value, Mapping)
        and value.get("contract_version") == GOVERNED_ROUTE_MANIFEST_VERSION
    ):
        from .adapter_contract import canonical_governed_route_manifest

        return canonical_governed_route_manifest(value)
    return canonical_governed_route_manifest_v2(value)

File: clients/typescript/src/operation-contract.ts
/** Additive public validators for `masugate.governed-route-manifest.v2`. */

import {
  GOVERNED_ROUTE_MANIFEST_VERSION,
  canonicalGovernedRouteManifest,
  canonicalJson,
  requireAdapterArgumentName,
  requireIdentifier,
  validateGovernedRouteManifest,
  type GovernedRouteManifest,
} from "./adapter-contract.js";
import type { JsonValue } from "./index.js";

export const GOVERNED_ROUTE_MANIFEST_V2_VERSION = "masugate.governed-route-manifest.v2";
export const DEFAULT_ROUTE_SCHEMA_CANONICAL_BYTES = 65_536;
export const DEFAULT_ROUTE_MANIFEST_CANONICAL_BYTES = 1_048_576;
export const MAX_GOVERNED_ROUTE_MANIFEST_ROUTES = 64;
export const MAX_ROUTE_ARTIFACT_FIELDS = 128;
export const MAX_ROUTE_CONNECTOR_CAPABILITIES = 64;

export interface GovernedRouteV2 {
  host_tool: string;
  action: string;
  input_schema: Record<string, JsonValue>;
  public_result_schema: Record<string, JsonValue>;
  artifact_fields: string[];
  owner:
    | { provider_id: string; position: "transactional" }
    | { provider_id: string; position: "protected-external"; connector_id: string };
  required_connector_capabilities: string[];
  maturity: "reference-effect" | "production-profile";
  compatibility: { route_manifest: typeof GOVERNED_ROUTE_MANIFEST_V2_VERSION; connector_contract: "masugate.connector.v1" };
}

export interface GovernedRouteManifestV2 {
  contract_version: typeof GOVERNED_ROUTE_MANIFEST_V2_VERSION;
  pack: { id: string; version: string; digest: string };
  routes: GovernedRouteV2[];
}

export type AnyGovernedRouteManifest = GovernedRouteManifest | GovernedRouteManifestV2;
const SECRET_MODEL_FIELD_PARTS = new Set([
  "credential", "credentials", "secret", "secrets", "token", "tokens", "password",
  "apikey", "privatekey", "accesskey",
]);
const SECRET_MODEL_FIELD_COMPOUNDS = new Set(["api_key", "private_key", "access_key"]);
const UTF8_ENCODER = new TextEncoder();

function modelField(value: unknown, field: string): string {
  if (typeof value !== "string") throw new TypeError(`${field} must be a model field`);
  const parsed = requireAdapterArgumentName(value);
  if (parsed === "runtime" || parsed.startsWith("model_")) {
    throw new TypeError(`${field} uses a reserved generated-host name`);
  }
  const parts = parsed.split("_");
  if (
    parts.some((part) => SECRET_MODEL_FIELD_PARTS.has(part)) ||
    parts.slice(0, -1).some((part, index) => SECRET_MODEL_FIELD_COMPOUNDS.has(`${part}_${parts[index + 1]}`))
  ) {
    throw new TypeError(`${field} cannot name secret or credential material`);
  }
  return parsed;
}

function denseArray(value: unknown, field: string): unknown[] {
  if (!Array.isArray(value)) throw new TypeError(`${field} must be an array`);
  for (let index = 0; index < value.length; index += 1) {
    if (!Object.hasOwn(value, index)) throw new TypeError(`${field} must not be sparse`);
  }
  return value;
}

function record(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}

function ownEntriesAtMost(
  value: Record<string, unknown>,
  maximum: number,
  field: string,
): [string, unknown][] {
  const entries: [string, unknown][] = [];
  for (const key in value) {
    if (!Object.hasOwn(value, key)) continue;
    if (entries.length === maximum) throw new TypeError(`${field} must contain at most ${maximum} entries`);
    entries.push([key, value[key]]);
  }
  return entries;
}

function exact(value: Record<string, unknown>, allowed: readonly string[], field: string): void {
  const actual = Object.keys(value).sort();
  const expected = [...allowed].sort();
  if (actual.length !== expected.length || actual.some((item, index) => item !== expected[index])) {
    throw new TypeError(`${field} must contain exactly: ${expected.join(", ")}`);
  }
}

function digest(value: unknown, field: string): string {
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) {
    throw new TypeError(`${field} must be a lowercase SHA-256 digest`);
  }
  return value;
}

function integer(value: unknown, field: string, minimum: number, maximum: number): number {
  if (
    typeof value !== "number" ||
    !Number.isInteger(value) ||
    value < minimum ||
    value > maximum
  ) {
    throw new TypeError(`${field} has invalid integer bounds`);
  }
  return value;
}

function schema(
  value: unknown,
  field: string,
  limit: number,
  objectRoot = false,
): Record<string, JsonValue> {
  let remaining = limit;
  const consume = (amount: number): void => {
    remaining -= amount;
    if (remaining < 0) throw new TypeError(`${field} canonical form exceeds configured limit`);
  };
  const canonicalBytes = (input: JsonValue): number => UTF8_ENCODER.encode(canonicalJson(input)).length;
  const visit = (raw: unknown, context: string, depth: number): Record<string, JsonValue> => {
    if (depth > 8) throw new TypeError(`${context} exceeds maximum schema nesting`);
    const node = record(raw, context);
    ownEntriesAtMost(node, 5, context);
    const kind = node["type"];
    if (kind === "object") {
      exact(node, ["type", "properties", "required", "additionalProperties"], context);
      if (node["additionalProperties"] !== false) throw new TypeError(`${context}.additionalProperties must be false`);
      const rawProperties = record(node["properties"], `${context}.properties`);
      const entries = ownEntriesAtMost(rawProperties, 128, `${context}.properties`);
      if (entries.length === 0 || entries.length > 128) throw new TypeError(`${context}.properties must contain 1 through 128 fields`);
      if (!Array.isArray(node["required"])) throw new TypeError(`${context}.required must be an array`);
      if (node["required"].length > entries.length) throw new TypeError(`${context}.required must name unique declared properties`);
      const propertyItems = entries.map(([name, child]) => [modelField(name, `${context}.properties key`), child] as const);
      const propertyNames = new Set(propertyItems.map(([name]) => name));
      const required: string[] = [];
      for (const [index, name] of node["required"].entries()) {
        if (typeof name !== "string") throw new TypeError(`${context}.required[${index}] must be a field`);
        const parsed = modelField(name, `${context}.required[${index}]`);
        if (!propertyNames.has(parsed) || required.includes(parsed)) {
          throw new TypeError(`${context}.required must name unique declared properties`);
        }
        required.push(parsed);
      }
      required.sort();
      consume(
        canonicalBytes({ type: "object", properties: {}, required, additionalProperties: false })
          + propertyItems.reduce((total, [name]) => total + canonicalBytes(name) + 1, 0)
          + propertyItems.length - 1,
      );
      const properties: Record<string, JsonValue> = {};
      for (const [name, child] of propertyItems) {
        properties[name] = visit(child, `${context}.properties.${name}`, depth + 1);
      }
      return { type: "object", properties, required, additionalProperties: false };
    }
    if (kind === "array") {
      const allowed = new Set(["type", "items", "minItems", "maxItems"]);
      if (Object.keys(node).some((key) => !allowed.has(key)) || !("items" in node) || !("maxItems" in node)) {
        throw new TypeError(`${context} array schemas must be explicitly bounded`);
      }
      const maximum = integer(node["maxItems"], `${context}.maxItems`, 0, 1024);
      const minimum = Object.hasOwn(node, "minItems")
        ? integer(node["minItems"], `${context}.minItems`, 0, maximum)
        : 0;
      consume(canonicalBytes({ type: "array", items: null, minItems: minimum, maxItems: maximum }) - "null".length);
      return { type: "array", items: visit(node["items"], `${context}.items`, depth + 1), minItems: minimum, maxItems: maximum };
    }
    if (kind === "string") {
      const allowed = new Set(["type", "minLength", "maxLength"]);
      if (Object.keys(node).some((key) => !allowed.has(key)) || !("maxLength" in node)) throw new TypeError(`${context} strings must declare maxLength`);
      const maximum = integer(node["maxLength"], `${context}.maxLength`, 0, 65_536);
      const minimum = Object.hasOwn(node, "minLength")
        ? integer(node["minLength"], `${context}.minLength`, 0, maximum)
        : 0;
      const parsed = { type: "string", minLength: minimum, maxLength: maximum } as const;
      consume(canonicalBytes(parsed));
      return parsed;
    }
    if (kind === "integer") {
      const allowed = new Set(["type", "minimum", "maximum"]);
      if (Object.keys(node).some((key) => !allowed.has(key))) throw new TypeError(`${context} integers permit only safe bounds`);
      const minimum = Object.hasOwn(node, "minimum")
        ? integer(node["minimum"], `${context}.minimum`, -9_007_199_254_740_991, 9_007_199_254_740_991)
        : -9_007_199_254_740_991;
      const maximum = Object.hasOwn(node, "maximum")
        ? integer(node["maximum"], `${context}.maximum`, minimum, 9_007_199_254_740_991)
        : 9_007_199_254_740_991;
      const parsed = { type: "integer", minimum, maximum } as const;
      consume(canonicalBytes(parsed));
      return parsed;
    }
    if (kind === "boolean" && Object.keys(node).length === 1) {
      const parsed = { type: "boolean" } as const;
      consume(canonicalBytes(parsed));
      return parsed;
    }
    throw new TypeError(`${context}.type is unsupported`);
  };
  const parsed = visit(value, field, 0);
  if (objectRoot && parsed["type"] !== "object") throw new TypeError(`${field} must be an object schema`);
  if (canonicalBytes(parsed) > limit) {
    throw new TypeError(`${field} canonical form exceeds configured limit`);
  }
  return parsed;
}

export function validateGovernedRouteManifestV2(
  value: unknown,
  options: { maxSchemaCanonicalBytes?: number; maxManifestCanonicalBytes?: number } = {},
): GovernedRouteManifestV2 {
  const limit = options.maxSchemaCanonicalBytes ?? DEFAULT_ROUTE_SCHEMA_CANONICAL_BYTES;
  if (!Number.isInteger(limit) || limit <= 0) throw new TypeError("schema canonical byte limit must be positive");
  const manifestLimit = options.maxManifestCanonicalBytes ?? DEFAULT_ROUTE_MANIFEST_CANONICAL_BYTES;
  if (!Number.isInteger(manifestLimit) || manifestLimit <= 0) throw new TypeError("route manifest canonical byte limit must be positive");
  const manifest = record(value, "governed route manifest v2");
  exact(manifest, ["contract_version", "pack", "routes"], "governed route manifest v2");
  if (manifest["contract_version"] !== GOVERNED_ROUTE_MANIFEST_V2_VERSION) throw new TypeError("governed route manifest v2.contract_version is unsupported");
  const rawPack = record(manifest["pack"], "governed route manifest v2.pack");
  exact(rawPack, ["id", "version", "digest"], "governed route manifest v2.pack");
  const pack = { id: requireIdentifier(rawPack["id"], "governed route manifest v2.pack.id"), version: requireIdentifier(rawPack["version"], "governed route manifest v2.pack.version"), digest: digest(rawPack["digest"], "governed route manifest v2.pack.digest") };
  const rawRoutes = denseArray(manifest["routes"], "governed route manifest v2.routes");
  if (rawRoutes.length === 0) throw new TypeError("governed route manifest v2.routes must be a non-empty array");
  if (rawRoutes.length > MAX_GOVERNED_ROUTE_MANIFEST_ROUTES) throw new TypeError(`governed route manifest v2.routes must contain at most ${MAX_GOVERNED_ROUTE_MANIFEST_ROUTES} entries`);
  const seenHostTools = new Set<string>();
  const seenActions = new Set<string>();
  const routes = rawRoutes.map((rawRoute, index) => {
    const field = `governed route manifest v2.routes[${index}]`;
    const route = record(rawRoute, field);
    exact(route, ["host_tool", "action", "input_schema", "public_result_schema", "artifact_fields", "owner", "required_connector_capabilities", "maturity", "compatibility"], field);
    const host_tool = requireIdentifier(route["host_tool"], `${field}.host_tool`);
    if (seenHostTools.has(host_tool)) throw new TypeError("governed route manifest v2.routes must not repeat host_tool");
    seenHostTools.add(host_tool);
    const action = requireIdentifier(route["action"], `${field}.action`, 255);
    if (seenActions.has(action)) throw new TypeError("governed route manifest v2.routes must not repeat action");
    seenActions.add(action);
    const ownerRaw = record(route["owner"], `${field}.owner`);
    let owner: GovernedRouteV2["owner"];
    if (ownerRaw["position"] === "transactional") {
      exact(ownerRaw, ["provider_id", "position"], `${field}.owner`);
      owner = { provider_id: requireIdentifier(ownerRaw["provider_id"], `${field}.owner.provider_id`), position: "transactional" };
    } else if (ownerRaw["position"] === "protected-external") {
      exact(ownerRaw, ["provider_id", "position", "connector_id"], `${field}.owner`);
      owner = { provider_id: requireIdentifier(ownerRaw["provider_id"], `${field}.owner.provider_id`), position: "protected-external", connector_id: requireIdentifier(ownerRaw["connector_id"], `${field}.owner.connector_id`) };
    } else throw new TypeError(`${field}.owner.position is invalid`);
    const artifactRaw = denseArray(route["artifact_fields"], `${field}.artifact_fields`);
    const capabilitiesRaw = denseArray(
      route["required_connector_capabilities"],
      `${field}.required_connector_capabilities`,
    );
    if (artifactRaw.length > MAX_ROUTE_ARTIFACT_FIELDS) throw new TypeError(`${field}.artifact_fields must contain at most ${MAX_ROUTE_ARTIFACT_FIELDS} entries`);
    if (capabilitiesRaw.length > MAX_ROUTE_CONNECTOR_CAPABILITIES) throw new TypeError(`${field}.required_connector_capabilities must contain at most ${MAX_ROUTE_CONNECTOR_CAPABILITIES} entries`);
    const artifact_fields = artifactRaw.map((name) => {
      if (typeof name !== "string") throw new TypeError(`${field}.artifact_fields must contain fields`);
      return modelField(name, `${field}.artifact_fields`);
    });
    const required_connector_capabilities = capabilitiesRaw.map((name) => requireIdentifier(name, `${field}.required_connector_capabilities`));
    if (new Set(artifact_fields).size !== artifact_fields.length || new Set(required_connector_capabilities).size !== required_connector_capabilities.length) throw new TypeError(`${field} arrays must not contain duplicates`);
    if (owner.position === "transactional" && required_connector_capabilities.length > 0) throw new TypeError(`${field} transactional route cannot require connector capabilities`);
    if (route["maturity"] !== "reference-effect" && route["maturity"] !== "production-profile") throw new TypeError(`${field}.maturity is invalid`);
    if (route["maturity"] === "production-profile" && owner.position !== "protected-external") throw new TypeError(`${field}.production-profile requires protected-external position`);
    const compatibility = record(route["compatibility"], `${field}.compatibility`);
    exact(compatibility, ["route_manifest", "connector_contract"], `${field}.compatibility`);
    if (compatibility["route_manifest"] !== GOVERNED_ROUTE_MANIFEST_V2_VERSION || compatibility["connector_contract"] !== "masugate.connector.v1") throw new TypeError(`${field}.compatibility is unsupported`);
    const input_schema = schema(route["input_schema"], `${field}.input_schema`, limit, true);
    const properties = input_schema["properties"] as Record<string, JsonValue>;
    if (artifact_fields.some((name) => !(name in properties))) throw new TypeError(`${field}.artifact_fields must name input properties`);
    if (artifact_fields.length > 0) {
      if (owner.position !== "protected-external") throw new TypeError(`${field}.artifact_fields require protected-external position`);
      const required = input_schema["required"] as string[];
      for (const artifact of artifact_fields) {
        const propertySchema = properties[artifact] as Record<string, JsonValue>;
        if (!required.includes(artifact) || propertySchema["type"] !== "string") {
          throw new TypeError(`${field}.artifact_fields must be required bounded string properties`);
        }
      }
    }
    return { host_tool, action, input_schema, public_result_schema: schema(route["public_result_schema"], `${field}.public_result_schema`, limit), artifact_fields, owner, required_connector_capabilities, maturity: route["maturity"], compatibility: { route_manifest: GOVERNED_ROUTE_MANIFEST_V2_VERSION, connector_contract: "masugate.connector.v1" } } as GovernedRouteV2;
  });
  routes.sort((left, right) => left.host_tool < right.host_tool ? -1 : left.host_tool > right.host_tool ? 1 : 0);
  const parsed: GovernedRouteManifestV2 = {
    contract_version: GOVERNED_ROUTE_MANIFEST_V2_VERSION,
    pack,
    routes,
  };
  if (new TextEncoder().encode(canonicalJson(parsed as unknown as JsonValue)).length > manifestLimit) {
    throw new TypeError("governed route manifest v2 canonical form exceeds configured limit");
  }
  return parsed;
}

export function validateAnyGovernedRouteManifest(value: unknown): AnyGovernedRouteManifest {
  if (record(value, "governed route manifest")["contract_version"] === GOVERNED_ROUTE_MANIFEST_VERSION) {
    return validateGovernedRouteManifest(value);
  }
  return validateGovernedRouteManifestV2(value);
}

export function canonicalGovernedRouteManifestV2(value: unknown): string {
  return canonicalJson(validateGovernedRouteManifestV2(value) as unknown as JsonValue);
}

export function canonicalAnyGovernedRouteManifest(value: unknown): string {
  if (record(value, "governed route manifest")["contract_version"] === GOVERNED_ROUTE_MANIFEST_VERSION) {
    return canonicalGovernedRouteManifest(value);
  }
  return canonicalGovernedRouteManifestV2(value);
}
